DIPs/DIP1NNN-CrazyPython.md

```diff
@@ -0,0 +1,167 @@
+| Status:         | Draft                                                           |
+
+## Abstract
+Often, it is necessary to allocate a data structure on the stack based on a
+size known at runtime. This proposal suggests adding dynamic-size static arrays,
+which are destroyed when they go out of scope. Dynamic-size static arrays and
+slices of them may be passed up the stack. The syntax is designed to encourage
+function signatures to use slices as much as possible, in contrast to the C
+convention of passing an array length and then a VLA. The overall goal is to
+help evolve D into a fewer-GC language.
+
+## Contents
+* [Rationale](#rationale)
+* [Prior Work](#prior-work)
+* [Description](#description)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Rationale
+`@nogc` programming is useful in many contexts. For example, currently WASM
+does not have a garbage collector, limiting D applications to `-betterC`
+mode. 
+
+It is often important to be able to quickly write `@nogc` code for
+performance-critical paths. After identifying hot functions using a profile,
+one should be able to alter a dynamic array that is only used within a single
+function to be allocated and deallocated automatically with `@nogc`.
+
+What is missing is an ergonomic way to dynamically allocate a static array on
+the stack built-in to D. You can take slices of static array as parameters
+
+```dlang
+/** A well-known algorithm for finding the topological sorting of a (possibly
+cyclic) directed graph. It deletes edges from its input as it runs, meaning
+adjacency_list is modified from within the function body.
+
+Because dynamic-size static arrays are value types, even when the function body
+modifies adjacency_list, the caller does not see the modifications.
+*/
+@nogc auto kahnsAlgorithm(int[][?] adjacency_list);
+```
+
+## Prior Work
+* `alloca`
+* `malloc` and `scope(exit) free`
+* Wait for D to support more C++ headers or use Calypso
+* Just use C++ 
+#### Use a dynamic array and copy the contents
+#### Templated code
+Make this legal:
+```dlang
+@nogc void example(N)(int[][N] adjacency_list) {
+return N;
+}
+unittest {
+int[][100] array;
+assert(example(array) == 100);
+}
+```
+Templated code is an alternative to having to hardcode the size into a function
+while keeping it `@nogc`. However, this would not work for arrays whose maximum
+size is unknown at runtime. 
+
+#### Copy parameters
+```dlang
+@nogc auto bfs(copy int[][] adjacency_list);
+```
+*Background:* It is already legal to receive a dynamic array in `@nogc` code.
+
+The `copy` parameter attribute creates a deep copy of the function parameter.
+In this case, the parameter `adjacency_list` is a static array of dynamic
+arrays. 
+
+When passing in a copy parameter, a copy is created at the entrance to the
+function body. Copy parameters are always implicitly `const copy`, since to the
+external caller, the passed in parameter is never modified.
```

Comment:
You cannot deep-copy about just about everything. Value types are fine, ref types (classes) are not.
And you can achieve the `@nogc`-ness by using `scope` parameter.